api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceAuthService.java

```diff
@@ -0,0 +1,222 @@
+    Map<String, WorkspaceAccessLevel> toAdd = new HashMap<>(updatedAclsMap);
+    for (Map.Entry<String, FirecloudWorkspaceAccessEntry> entry : aclsMap.entrySet()) {
+      String currentUserEmail = entry.getKey();
+      WorkspaceAccessLevel updatedAccess = toAdd.get(currentUserEmail);
+      if (updatedAccess != null) {
+        FirecloudWorkspaceACLUpdate currentUpdate = new FirecloudWorkspaceACLUpdate();
+        currentUpdate.setEmail(currentUserEmail);
+        currentUpdate =
+            WorkspaceAuthService.updateFirecloudAclsOnUser(updatedAccess, currentUpdate);
+        updateACLRequestList.add(currentUpdate);
+        toAdd.remove(currentUserEmail);
+      } else {
+        // This is how to remove a user from the FireCloud ACL:
+        // Pass along an update request with NO ACCESS as the given access level.
+        // Note: do not do groups.  Unpublish will pass the specific NO_ACCESS acl
+        // TODO [jacmrob] : have all users pass NO_ACCESS explicitly? Handle filtering on frontend?
+        if (!currentUserEmail.equals(registeredUsersGroup)) {
+          FirecloudWorkspaceACLUpdate removedUser = new FirecloudWorkspaceACLUpdate();
+          removedUser.setEmail(currentUserEmail);
+          removedUser =
+              WorkspaceAuthService.updateFirecloudAclsOnUser(
+                  WorkspaceAccessLevel.NO_ACCESS, removedUser);
+          updateACLRequestList.add(removedUser);
+        }
+      }
+    }
+
+    // Iterate through remaining new roles; add them
+    for (Map.Entry<String, WorkspaceAccessLevel> remainingRole : toAdd.entrySet()) {
+      FirecloudWorkspaceACLUpdate newUser = new FirecloudWorkspaceACLUpdate();
+      newUser.setEmail(remainingRole.getKey());
+      newUser = WorkspaceAuthService.updateFirecloudAclsOnUser(remainingRole.getValue(), newUser);
+      updateACLRequestList.add(newUser);
+    }
+    FirecloudWorkspaceACLUpdateResponseList fireCloudResponse =
+        fireCloudService.updateWorkspaceACL(
+            workspace.getWorkspaceNamespace(), workspace.getFirecloudName(), updateACLRequestList);
+    if (fireCloudResponse.getUsersNotFound().size() != 0) {
+      String usersNotFound = "";
+      for (int i = 0; i < fireCloudResponse.getUsersNotFound().size(); i++) {
+        if (i > 0) {
+          usersNotFound += ", ";
+        }
+        usersNotFound += fireCloudResponse.getUsersNotFound().get(i).getEmail();
+      }
+      throw new BadRequestException(usersNotFound);
+    }
+
+    // Finally, keep OWNER and billing project users in lock-step. In Rawls, OWNER does not grant
+    // canCompute on the workspace / billing project, nor does it grant the ability to grant
+    // canCompute to other users. See RW-3009 for details.
+    for (String email : Sets.union(updatedAclsMap.keySet(), aclsMap.keySet())) {
+      String fromAccess =
+          aclsMap
+              .getOrDefault(email, new FirecloudWorkspaceAccessEntry().accessLevel(""))
```

Comment:
Should this be WorkspaceAccessLevel.NO_ACCESS? instead of ""

Reply:
Looks like it's only being compared to "OWNER" so this won't have a functional change but "NO ACCESS" makes more sense than "" to me.